gson/src/main/java/com/google/gson/GsonBuilder.java

```diff
@@ -664,14 +666,16 @@ public GsonBuilder setDateFormat(int dateStyle, int timeStyle) {
    * @param typeAdapter This object must implement at least one of the {@link TypeAdapter},
    * {@link InstanceCreator}, {@link JsonSerializer}, and a {@link JsonDeserializer} interfaces.
    * @return a reference to this {@code GsonBuilder} object to fulfill the "Builder" pattern
+   * @throws IllegalArgumentException if the Type adapter being registered is for Object class or (JsonElements or any of its subclasses)
    */
   @CanIgnoreReturnValue
   public GsonBuilder registerTypeAdapter(Type type, Object typeAdapter) {
     Objects.requireNonNull(type);
-    $Gson$Preconditions.checkArgument(typeAdapter instanceof JsonSerializer<?>
+    $Gson$Preconditions.checkArgument((typeAdapter instanceof JsonSerializer<?>
         || typeAdapter instanceof JsonDeserializer<?>
         || typeAdapter instanceof InstanceCreator<?>
-        || typeAdapter instanceof TypeAdapter<?>);
+        || typeAdapter instanceof TypeAdapter<?>)
+        && !isTypeObjectOrJsonElements(type));
```

Comment:
@Marcono1234 , Although I liked (and prefer) to provide a useful exception message, this breaks the new condition out of `$Gson$Preconditions.checkArgument` , which may set a trend for future or existing checks... 
We could clarify this in Java-doc or usage guide and still provide a generic IllegalArgumentException, what do you think?

Reply:
> this breaks the new condition out of `$Gson$Preconditions.checkArgument` , which may set a trend for future or existing checks...

But is this (the trend of providing more helpful exception messages) an issue?

The current `checkArgument` covers only validation of `typeAdapter`, I think it should either way not be extended to also validate `type`. Otherwise the name "checkArgument" (instead of plural "checkArgument<ins>s</ins>") would also be misleading. And here separate calls to `checkArgument` are used as well:
https://github.com/google/gson/blob/ad5cd31498a7d6835957b7fe3a9ece4b12753867/gson/src/main/java/com/google/gson/internal/%24Gson%24Types.java#L605-L606

And if we then have a separate validation call anyway, then I think it would be useful to provide a specific exception message. This will especially be useful for users who might experience this exception after upgrading their Gson dependency, and where a `IllegalArgumentException` without any message would not be very helpful.

Reply:
That makes sense, I am thinking to override the checkArgument with an error message for future use and current use as follows:

```
$Gson$Preconditions.checkArgument(isTypeObjectOrJsonElement(type), "Cannot override built-in adapter for " + type);

  public static void checkArgument(boolean condition, String message) {
    if (!condition) {
      throw new IllegalArgumentException(message);
    }
  }
```

Reply:
Sounds good to me. But in case you want to adjust all `checkArgument` calls in the Gson code to provide a message, please keep the existing `checkArgument(boolean)` nonetheless and mark it deprecated, similar to `$Gson$Preconditions.checkNotNull`. Because it looks like unfortunately [some projects are using this internal Gson method](https://github.com/search?q=%22%24Gson%24Preconditions.checkArgument%22&type=code) (though a lot of these search result are in copies of the Gson code).

Reply:
Please _don't_ make unrelated changes in this PR. It might make sense to change all `checkArgument` calls so they have a message, though there are code-size implications. But if we are to consider such a change it should be in a separate PR.

Reply:
Sorry, please ignore then what I said about `checkArgument`.

But @eamonnmcmanus for these `GsonBuilder` methods, throwing an `IllegalArgumentException` with message (constructed manually, without `checkArgument`) as proposed by me originally in this thread should be fine then, right? Since this is directly related to these changes.

Reply:
ok I am going with original proposed change to add a specific check outside of the `checkArgument()` method.